addons/mail/static/src/model/model_manager.js

```diff
@@ -431,25 +430,49 @@ export class ModelManager {
     }
 
     /**
+     * Adds fields, methods, and other stuff from the model definition to the
+     * Model.
+     *
      * @private
-     * @param {mail.model} Model class
-     * @param {Object} patch
+     * @param {mail.model} Model
      */
-    _applyModelPatchFields(Model, patch) {
-        for (const [fieldName, field] of Object.entries(patch)) {
-            Model.fields[fieldName] = field;
+    _applyModelDefinition(Model) {
+        const definition = registry[Model.name];
+        Object.assign(Model, definition.modelMethods);
+        Object.assign(Model.prototype, definition.recordMethods);
+        // use `modelName` rather than `name` to avoid conflicting with the name
+        // inherited from Function
+        Model.modelName = definition.name;
+        // Make model manager accessible from Model.
+        Model.modelManager = this;
+        Model.fields = {};
+        /**
+        * Contains all records. key is local id, while value is the record.
+        */
+        Model.__records = {};
+        Model.identifyingFields = definition.identifyingFields;
+        Model.__identifyingFields = Model.identifyingFields;
+        const identifyingFieldsFlattened = new Set();
+        for (const identifyingElement of Model.identifyingFields) {
+            const identifyingFields = typeof identifyingElement === 'string'
+                ? [identifyingElement]
+                : identifyingElement;
+            for (const identifyingField of identifyingFields) {
+                identifyingFieldsFlattened.add(identifyingField);
+            }
         }
+        Model.__identifyingFieldsFlattened = identifyingFieldsFlattened;
+        this._listenersObservingAllByModel.set(Model, new Map());
+        this.models[Model.modelName] = Model;
     }
 
     /**
      * @private
-     * @param {Object} Models
      * @throws {Error} in case some declared fields are not correct.
      */
-    _checkDeclaredFieldsOnModels(Models) {
-        for (const Model of Object.values(Models)) {
-            for (const fieldName in Model.fields) {
-                const field = Model.fields[fieldName];
+    _checkDeclaredFieldsOnModels() {
+        for (const Model of Object.values(this.models)) {
```

Comment:
I renamed it `_checkFieldsOnGeneratedModels` for now. But I wonder if we shouldn't move all this stuff to the model core. Since patches are no longer applied during model generation, I see no reason not to do so as soon as the model is registered. It would make more sense to me.

Reply:
Let's keep existing names for now then, as we are unsure. (new name is not really better)